### DISCUSSION TITLE
How to sort tree by node attribute?

### DISCUSSION THREAD

**Discussion by Otterpatsch** [MunifTanjim/nui.nvim#341] [Q&A]
Hello there,
I want to sort the nodes by their attribute timestemp. But still keep the parent/child relationship intact. So basically only sort the root nodes even tho the child nodes should be sorted too but thats optional if not feasible. 

So currently  i guess as i use pairs to add nodes  to the tree is unsorted as values was sorted
```lua
	local node_by_id = M.values_to_nodes(values)

	repo.comment_tree = NuiTree({
		bufnr = comment_split.bufnr,
		get_node_id = function(node)
			-- this is telling NuiTree where we're storing the id
			return node.id
		end,
		prepare_node = function(node) end,
	})

	for id in pairs(node_by_id) do
		M.add_node_to_tree(id, node_by_id)
	end
```
As the keys of node_by_id is some integer as string (id attribute of a value from a api call)
I thought okay sort afterwards the nodes of the tree and just create a new tree with the sorted nodes. Which atleast doenst throw any error but is also not sorting as expected.

as the output is lacking some nodes and the parent/child relationship is broken. 

###I tried: 

```lua
	local nodes_sorted = repo.comment_tree:get_nodes()
	table.sort(nodes_sorted, function(a, b)
		if not a:get_parent_id() then
			return a.timestemp > b.timestemp
		end
		return false
	end)

	repo.comment_tree = NuiTree({
		bufnr = comment_split.bufnr,
		get_node_id = function(node)
			-- this is telling NuiTree where we're storing the id
			return node.id
		end,
		prepare_node = function(node)
			local parent_node = node_by_id[node:get_parent_id()]
			return M.node_visualize(node, parent_node)
		end,
		nodes = nodes_sorted,
	})

```
### bad result
 ![image](https://github.com/MunifTanjim/nui.nvim/assets/29692215/b2f7d62f-5866-46e1-9d96-445fe828fb7c)

### unsorted result 
![image](https://github.com/MunifTanjim/nui.nvim/assets/29692215/29b525aa-ad92-46ba-accd-a927a5bfd7c6)

#### Things which might be helpful to reproduce 
repo [function M.comments_view(values)](https://github.com/Otterpatsch/bitbucket.nvim/blob/feature/add-discussion-section/lua/bitbucket/comments/tree.lua#L31)

values is basically just a concataned table of all response.bodys of those 3 pages
```zsh
curl --request GET \
 --url 'https://api.bitbucket.org/2.0/repositories//otterpatsch/bitbucket.nvim.example/pullrequests/1/comments?page=1' \
 --header 'Accept: application/json' \
-o comments1.json

curl --request GET \
 --url 'https://api.bitbucket.org/2.0/repositories//otterpatsch/bitbucket.nvim.example/pullrequests/1/comments?page=2' \
 --header 'Accept: application/json' \
-o comments2.json

curl --request GET \
 --url 'https://api.bitbucket.org/2.0/repositories//otterpatsch/bitbucket.nvim.example/pullrequests/1/comments?page=3' \
 --header 'Accept: application/json' \
-o comments3.json
```

**Comment by MunifTanjim** [comment]
`pairs` in Lua doesn't guarantee any order.

You can sort `values` first, then iterated over sorted `values` with `ipairs` instead of doing `pairs(node_by_id)`.

Currently `NuiTree` doesn't provide any API to sort the nodes... it will show nodes in the order you added them. So you need to sort the items first and then add them as nodes.
